src/basic/process-util.c

```diff
@@ -158,25 +157,18 @@ int get_process_cmdline(pid_t pid, size_t max_columns, ProcessCmdlineFlags flags
         if ((size_t) 4 * max_columns + 1 < max_columns)
                 max_length = MIN(max_length, (size_t) 4 * max_columns + 1);
 
-        t = new(char, max_length);
-        if (!t)
-                return -ENOMEM;
-
-        k = fread(t, 1, max_length, f);
         if (k > 0) {
                 /* Arguments are separated by NULs. Let's replace those with spaces. */
                 for (size_t i = 0; i < k - 1; i++)
                         if (t[i] == '\0')
                                 t[i] = ' ';
 
                 t[k - 1] = '\0'; /* Normally, t[k - 1] is already NUL, so this is just a guard in case of short read */
-        } else {
-                /* We only treat getting nothing as an error. We *could* also get an error after reading some
-                 * data, but we ignore that case, as such an error is rather unlikely and we prefer to get
-                 * some data rather than none. */
-                if (ferror(f))
-                        return -errno;
 
+                /* Truncate with NUL if it's longer than what we want. String will be reallocated before return. */
+                if (k > max_length)
+                        t[max_length - 1] = '\0';
```

Comment:
this should be unnecessary. escape_non_printable_full() will already truncate as needed. I figure you can drop `max_length` altogether when using read_full_virtual_file(), as that enforces an internal maximum size anyway